### PR TITLE
replace FLnext linked list with Barray

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by ibuclaw** [comment]
Something is not quite right still:
```
src/dmd/backend/gloop.d:71:5: error: dmd.backend.gloop.loop.~this at src/dmd/backend/gloop.d:71:5 conflicts with dmd.backend.gloop.loop.~this at src/dmd/backend/gloop.d:71:5
```

**Comment by ibuclaw** [comment]
Seen on FreeBSD and Win32 in the form of linker errors.
```
ERROR: ../generated/freebsd/release/32/backend.o: In function `loop::~loop()':
/usr/home/braddr/sandbox/at-client/pull-4051625-FreeBSD_32/dmd/src/dmd/backend/aarray.d:(.text._ZN4loopD1Ev+0x108): multiple definition of `loop::~loop()'
../generated/freebsd/release/32/backend.o:/usr/home/braddr/sandbox/at-client/pull-4051625-FreeBSD_32/dmd/src/dmd/backend/aarray.d:(.text._ZN4loopD1Ev+0x0): first defined here
/usr/bin/ld: Warning: size of symbol `_ZN4loopD1Ev' changed from 262 in ../generated/freebsd/release/32/backend.o to 137 in ../generated/freebsd/release/32/backend.o
../generated/freebsd/release/32/backend.o: In function `loop::~loop()':
/usr/home/braddr/sandbox/at-client/pull-4051625-FreeBSD_32/dmd/src/dmd/backend/aarray.d:(.text._ZN4loopD1Ev+0x198): multiple definition of `loop::~loop()'
../generated/freebsd/release/32/backend.o:/usr/home/braddr/sandbox/at-client/pull-4051625-FreeBSD_32/dmd/src/dmd/backend/aarray.d:(.text._ZN4loopD1Ev+0x0): first defined here
/usr/bin/ld: Warning: size of symbol `_ZN4loopD1Ev' changed from 137 in ../generated/freebsd/release/32/backend.o to 371 in ../generated/freebsd/release/32/backend.o
../generated/freebsd/release/32/backend.o: In function `Iv::~Iv()':
/usr/home/braddr/sandbox/at-client/pull-4051625-FreeBSD_32/dmd/src/dmd/backend/aarray.d:(.text._ZN2IvD1Ev+0x60): multiple definition of `Iv::~Iv()'
../generated/freebsd/release/32/backend.o:/usr/home/braddr/sandbox/at-client/pull-4051625-FreeBSD_32/dmd/src/dmd/backend/aarray.d:(.text._ZN2IvD1Ev+0x0): first defined here
/usr/bin/ld: Warning: size of symbol `_ZN2IvD1Ev' changed from 91 in ../generated/freebsd/release/32/backend.o to 53 in ../generated/freebsd/release/32/backend.o
../generated/freebsd/release/32/backend.o: In function `Iv::~Iv()':
/usr/home/braddr/sandbox/at-client/pull-4051625-FreeBSD_32/dmd/src/dmd/backend/aarray.d:(.text._ZN2IvD1Ev+0x98): multiple definition of `Iv::~Iv()'
../generated/freebsd/release/32/backend.o:/usr/home/braddr/sandbox/at-client/pull-4051625-FreeBSD_32/dmd/src/dmd/backend/aarray.d:(.text._ZN2IvD1Ev+0x0): first defined here
/usr/bin/ld: Warning: size of symbol `_ZN2IvD1Ev' changed from 53 in ../generated/freebsd/release/32/backend.o to 128 in ../generated/freebsd/release/32/backend.o
cc: error: linker command failed with exit code 1 (use -v to see invocation)
Error: linker exited with status 1
```

**Comment by ibuclaw** [comment]
Huh, looks like C++ dtors have never worked.

File: src/dmd/backend/gloop.d [hunk 1]
```diff
@@ -166,7 +159,15 @@ struct Iv
...
     elem **IVincr;          // pointer to parent of IV increment elem
-    famlist *IVfamily;      // variables in this family
+    Barray!famlist IVfamily;      // variables in this family
+
+    ~this()
+    {
+        foreach (ref fl; IVfamily)
+        {
+            fl.__dtor();
+        }
```
Comment (ibuclaw) [inline src/dmd/backend/gloop.d:169]: The "fix" would be to move this to `Barray`.
```
static if (__traits(compiles, T.__dtor))
{
    foreach (ref e; array)
        e.__dtor();
}
```

**Review by ibuclaw — DISMISSED** [review]

File: src/dmd/backend/gloop.d (hunk 1 above)
Comment (ibuclaw) [reply src/dmd/backend/gloop.d:169]: Edited to use `__xdtor` instead.  Other places where you explicitly call `__dtor` on each element before calling `array.__dtor()` can be removed as well.

**Comment by WalterBright** [comment]
Another way is to put Barray back where I originally had it, with no destructor. It was never designed to work with destructors, as this mess illustrates.

**Comment by WalterBright** [comment]
One of the things Barray is for is to re-use its array, and this just doesn't fit right with destructors.

**Review by ibuclaw — APPROVED** [review]

**Pull request merged**